Linopy Integration

GbotemiB opened the pull request:
# Closes # (if applicable).

## Changes proposed in this Pull Request
Integrate linopy following https://github.com/PyPSA/pypsa-eur/pull/625 and https://github.com/pypsa-meets-earth/pypsa-earth/pull/796

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [ ] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

GbotemiB committed:
modified files to adjust to linopy

GbotemiB committed:
removed unused comments

GbotemiB committed:
added config to include new parameters

ekatef commented:
Hello @GbotemiB!

Thanks a lot for opening the PR! Currently, CI is failing, and it looks like the reason is the recent environment troubles we experienced in PyPSA-Earth. Those troubles have been fixed recently with [this PR](https://github.com/pypsa-meets-earth/pypsa-earth/pull/1047) and merged into Linopy PR in PyPSA-Earth. Could you please change the update the commit on which your PR relays? I think, [this one](<https://github.com/pypsa-meets-earth/pypsa-earth/pull/796/commits/cfb251fc20e4b5a392b077f15f551fb282ed5015>) should work fine.

GbotemiB commented:
Hi @ekatef,

Thanks for looking into this. I will review the changes and revert.

ekatef commented:
> Hi @ekatef,
> 
> Thanks for looking into this. I will review the changes and revert.

Hey @GbotemiB, no worries and no need to revert! 🙂 We just need to include the recent changes form PyPSA-Earth upstream.

GbotemiB committed:
config tailored to accomodate new functionalities

GbotemiB committed:
merge upstream for pypsa-earth

ekatef commented on scripts/solve_network.py:
```diff
@@ -314,78 +319,72 @@ def add_chp_constraints(n):
     co2_stores = n.stores.loc[n.stores.carrier == "co2 stored"].index
 
-    if co2_stores.empty or ("Store", "e") not in n.variables.index:
+    if co2_stores.empty:  # or ("Store", "e") not in n.variables.index:
```
Just for better understanding: what has been the reason of commenting-out the second part of the condiiton?

GbotemiB replied:
The variables method no longer exists in the network, which is why I commented it out. I also noticed something in pypsa-eur implementation.

GbotemiB replied:
There seems to be no `variable` method in `n`. I think it relates to the new data structures implemented in `pypsa.Network`

I noticed the same similarities [here](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796/commits/cfb251fc20e4b5a392b077f15f551fb282ed5015) and in pypsa-eur.

ekatef replied:
You mean `n.variables` component, right? 

Agree that some adjustment may be needed, but we have to avoid losses in functionality. It would be great to understand the purpose of having `("Store", "e") not in n.variables.index` condition here, check if it's still needed and find a way to re-implement it. May be a good point to discuss with the cross-sectoral team.

GbotemiB replied:
I don't know the reason or the functionality. But, I checked similar implementations in pypsa-eur, before the linopy integration, and after the integration, and it seems to have been done in the same way.

FabianHofmann replied:
you can access the variable with `n.model.variables`. so the equivalent check would be `"Store-e" not in n.model.variables`

ekatef commented on scripts/solve_network.py:
```diff
@@ -477,37 +476,47 @@ def extra_functionality(n, snapshots):
+
+    if kwargs["solver_name"] == "gurobi":
+        logging.getLogger("gurobipy").setLevel(logging.CRITICAL)
+    rolling_horizon = cf_solving.pop("rolling_horizon", False)
```
Not sure we have an option for the rolling horizon... Agree that it may be interesting to advance implementation, but it feels like an issue per se. What do you think @GbotemiB?

GbotemiB replied:
Yeah, we currently don't have an implementation for rolling horizon. But, I picked it from pypsa-eur. We can probably remove it for now and include it for advance implementation.

GbotemiB replied:
Yeah. We don't have an implementation for `rolling horizon`. We can simply just leave it out and include it for an advance implementation.

ekatef commented:
Thanks a lot @GbotemiB! The update has worked 🎉 
CI is failing currently but the reason is that some remote source is unavailable (`urllib.error.HTTPError: HTTP Error 503: Service Unavailable`).

The PR itself looks great! Is it ready for review?

GbotemiB commented:
@ekatef, the PR is ready for review. I didn't mark it for review because the [Linopy PR](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796/commits/cfb251fc20e4b5a392b077f15f551fb282ed5015) in pypsa-earth is still in draft.

I have investigated the error in the CI. It relates to `prepare_transport_data_input:download_number_of_vehicles.py` which is supposed to access this link [here](https://apps.who.int/gho/athena/data/GHO/RS_194?filter=COUNTRY:*&ead=&x-sideaxis=COUNTRY;YEAR;DATASOURCE&x-topaxis=GHO&profile=crosstable&format=csv), but at the moment the site is currently not reachable. This should be the reason for the error.

GbotemiB committed:
removed rolling horizon option

davide-f commented on scripts/solve_network.py:
```diff
@@ -314,78 +319,72 @@ def add_chp_constraints(n):
-            (1, link_p[heat_ext]),
-            (1, link_p[electric_ext].values),
-            (-1, link_p_nom[electric_ext].values),
+        lhs = (
+            (1 * link_p.loc[heat_ext]),
+            (1 * link_p.loc[electric_ext]),
+            (1 * link_p_nom.loc[electric_ext]),
         )
```
This one cannot work.
It feels to me that probably we need to slightly change the CI options to enable more of these functions to actually be tested.
Moreover, pay attention on signs, it seems something has went wrong here

GbotemiB replied:
Thanks @davide-f for catching this. I will revise the implementation.

davide-f reviewed (commented):
Hello :D
Just adding few cents :)
Great @GbotemiB :D

GbotemiB commented on config.default.yaml:
```diff
@@ -371,17 +371,65 @@ solving:
+    glpk-default: {} # Used in CI
 
   mem: 30000 #memory in MB; 20 GB enough for 50+B+I+H2; 100 GB for 181+B+I+H2
+  walltime: "12:00:00"
```
Thanks @davide-f, I will check it out.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

GbotemiB committed:
merge from ekatef dev-linopy branch

GbotemiB committed:
update config variable

GbotemiB committed:
updates signs

ekatef commented on scripts/solve_network.py:
```diff
@@ -314,78 +319,72 @@ def add_chp_constraints(n):
         return
 
-    vars_final_co2_stored = get_var(n, "Store", "e").loc[sns[-1], co2_stores]
+    vars_final_co2_stored = n.model["Store-e"].loc[sns[-1], co2_stores]
```
As a general comment, indexing in `linopy` may be a bit delicate. My observation is that there may be no one-to-one correspondence with old functions. Some testing may be needed, especially there is no exact match with PyPSA-Eur.

GbotemiB replied:
Do you have any ideas on how we can test the functionalities?

from my end, I think we can test the original functionalities alongside the new ones, and compare the results.

ekatef commented on scripts/solve_network.py:
```diff
@@ -314,78 +319,72 @@ def add_chp_constraints(n):
+    vars_final_co2_stored = n.model["Store-e"].loc[sns[-1], co2_stores]
 
-    lhs = linexpr((1, vars_final_co2_stored)).sum()
+    lhs = (1 * vars_final_co2_stored).sum()
```
Not sure we need to multiply by `1` here and in all similar fragments. 

My understanding is that a linopy equivalent of `linexpr((1, vars_final_co2_stored)).sum()` is `vars_final_co2_stored.sum()`. But that is not necessarily 100% correct understanding, and happy to discuss. Do you have any ideas on that? Have you tested these parts?

GbotemiB replied:
You are right. The `1` multiple is redundant. I did a bit of debugging to verify the expression already. But I am open to more ways to test these implementation.

ekatef commented:
Hello @GbotemiB!

Very nice and quite comprehensive work! Thank you for addressing the review comments. I think, now we have a good basic linopy-based implementation both for pypsa-earth and pypsa-earth-sec. The next step must be a good testing, as `linopy` is quite a powerful tool which should be used in a mindful way. Along with PyPSA-Eur implementation, there are [great examples](https://github.com/PyPSA/linopy/blob/master/examples) on linopy usage which can be very helpful to get the main principles. I have also collected some observations on `linopy` syntax which may be relevant for our PRs: [see this comment](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796#issuecomment-2200080508). 

Absolutely agree with @davide-f that CI doesn't catch the points which are crucial for functionality of both `linopy` PRs. Which mean that we can't fully relate on CI in this case, and some a manual testing is necessary. My feeling is that the following tests are needed:
1. Execute some extra-functionality functions. In particular, `add_chp_constraints(n)` looks quite complicated, and it's local testing is highly desirable.
2. Check that the objective didn't change due to this PR. Or didn't change significantly, at the very least :)
3. Backward compatibility is respected: PR runs also with the old environment.

@GbotemiB does the plan sound reasonable for you?

GbotemiB commented:
> 1. Execute some extra-functionality functions. In particular, add_chp_constraints(n) looks quite complicated, and it's local testing is highly desirable.
> 2. Check that the objective didn't change due to this PR. Or didn't change significantly, at the very least :)
> 3. Backward compatibility is respected: PR runs also with the old environment.
> @GbotemiB does the plan sound reasonable for you?

@ekatef Sounds perfect. I think in terms of priority, `1` should come ontop.

GbotemiB commented:
> * Check that the objective didn't change due to this PR. Or didn't change significantly, at the very least :)



> > 1. Execute some extra-functionality functions. In particular, add_chp_constraints(n) looks quite complicated, and it's local testing is highly desirable.
> > 2. Check that the objective didn't change due to this PR. Or didn't change significantly, at the very least :)
> > 3. Backward compatibility is respected: PR runs also with the old environment.
> >    @GbotemiB does the plan sound reasonable for you?
> 
> @ekatef Sounds perfect. I think in terms of priority, `1` should come on top.

I should also mention here that in the original pypsa-earth-sec, the test config results in `nan` as objective value. I think this is one of the things that CI doesn't catch.

GbotemiB committed:
modified chp constraints

GbotemiB commented:
> Hello @GbotemiB!
> 
> Very nice and quite comprehensive work! Thank you for addressing the review comments. I think, now we have a good basic linopy-based implementation both for pypsa-earth and pypsa-earth-sec. The next step must be a good testing, as `linopy` is quite a powerful tool which should be used in a mindful way. Along with PyPSA-Eur implementation, there are [great examples](https://github.com/PyPSA/linopy/blob/master/examples) on linopy usage which can be very helpful to get the main principles. I have also collected some observations on `linopy` syntax which may be relevant for our PRs: [see this comment](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796#issuecomment-2200080508).
> 
> Absolutely agree with @davide-f that CI doesn't catch the points which are crucial for functionality of both `linopy` PRs. Which mean that we can't fully relate on CI in this case, and some a manual testing is necessary. My feeling is that the following tests are needed:
> 
> 1. Execute some extra-functionality functions. In particular, `add_chp_constraints(n)` looks quite complicated, and it's local testing is highly desirable.
> 2. Check that the objective didn't change due to this PR. Or didn't change significantly, at the very least :)
> 3. Backward compatibility is respected: PR runs also with the old environment.
> 
> @GbotemiB does the plan sound reasonable for you?

1. About testing the implementation of `add_chp_constraints(n)`, I confirmed that the implementation Is now just as it is in pypsa-eur. I havent found a way to test it, I have a plan to run the script in debug mode and test it that way.
2. I have confirmed that the objective didn't change due to this PR. I ran the CI locally which returns a `nan` for the objective value. I am already in the process of investigating why this happened with @finozzifa.
3. The current PR can't run with the old environment because of the latest version of pypsa.

GbotemiB committed:
precommit fixes

GbotemiB committed:
add_chp_constraints fixed and tested

GbotemiB committed:
added chp to extra_functionalities

davide-f reviewed (commented):
Few comments; great @GbotemiB! :D
As a comment, do you feel like by simply changing few parameters in the config test yaml, the functions you modified could be included in the CI?
That may ease the process (if we also fix the CI, but that may not be done here)

GbotemiB committed:
reset snakemake default

GbotemiB committed:
remove unused import

GbotemiB committed:
Merge remote-tracking branch 'origin/main' into linopy-dev

GbotemiB commented on Snakefile:
```diff
@@ -189,7 +188,7 @@ rule prepare_sector_network:
         nodal_transport_data="resources/demand/nodal_transport_data_{demand}_s{simpl}_{clusters}_{planning_horizons}.csv",
         overrides="data/override_component_attrs",
-        clustered_pop_layout="resources/population_shares/pop_layout_elec_s{simpl}_{clusters}_{planning_horizons}.csv",
+        clustered_pop_layout="resources/population_shares/pop_layout_elec_s{simpl}_{clusters}.csv",
```
Thanks for catching this. I have resolved the merge conflict.

GbotemiB commented:
> Few comments; great @GbotemiB! :D As a comment, do you feel like by simply changing few parameters in the config test yaml, the functions you modified could be included in the CI? That may ease the process (if we also fix the CI, but that may not be done here)

Hi @davide-f, could you please clarify what you mean here?
Do you mean making changes to the config test file?

davide-f commented:
> Hi @davide-f, could you please clarify what you mean here? Do you mean making changes to the config test file?

Some functions in the solve_network are triggered by options in the config file, some of them may be off by default which means that they are not tested.
Not a priority, but worth mentioning if it may help the revision

GbotemiB commented:
> > Hi @davide-f, could you please clarify what you mean here? Do you mean making changes to the config test file?
> 
> Some functions in the solve_network are triggered by options in the config file, some of them may be off by default which means that they are not tested. Not a priority, but worth mentioning if it may help the revision

The option to test the `add_chp_contraints` in the test config is already set to true. The issue is that the solve_network script doesn't call the function. I have included it in my PR.

GbotemiB committed:
Merge remote-tracking branch 'origin/main' into linopy-dev

GbotemiB committed:
update from upstream

GbotemiB commented:
Just adding here the results from the linopy run

```bash
INFO:__main__:Objective function: 7105181487.897996
INFO:__main__:Objective constant: 142949224.50470072
```

Results from main default.
```bash
INFO:__main__:Objective function: 7993595361.0
INFO:__main__:Objective constant: 142949224.50470072
```

~~The percentage difference is about 1%.~~

ekatef commented:
> Just adding here the results from the linopy run
> 
> ```shell
> INFO:__main__:Objective function: 7105181487.897996
> INFO:__main__:Objective constant: 142949224.50470072
> ```

Hello @GbotemiB, great to see that you have managed to fix the NaN issue. Could you please also post the objective values for the run from main?

ekatef commented:
> > Just adding here the results from the linopy run
> > ```shell
> > INFO:__main__:Objective function: 7105181487.897996
> > INFO:__main__:Objective constant: 142949224.50470072
> > ```
> 
> Hello @GbotemiB, great to see that you have managed to fix the NaN issue. Could you please also post the objective values for the run from main with the implemented NaN fix?

davide-f commented:
Regarding the yamlint issue about brackets {} it is possible to fix by changing the file .yamlint by adding the entry:
``` yaml
rules:
  empty-values:
     forbid-in-flow-mappings: false
```

For more: https://github.com/suinplayground/kubernetes-playground/blob/main/.yamllint.yml

GbotemiB commented:
> Regarding the yamlint issue about brackets {} it is possible to fix by changing the file .yamlint by adding the entry:
> 
> ```yaml
> rules:
>   empty-values:
>      forbid-in-flow-mappings: false
> ```
> 
> For more: https://github.com/suinplayground/kubernetes-playground/blob/main/.yamllint.yml

It might be better to just handle the issues with  the config file in a separate PR. Since this is an issue pertaining to both sec and power model, I can go ahead to open a PR in earth model.

GbotemiB commented:
> Just adding here the results from the linopy run
> 
> ```shell
> INFO:__main__:Objective function: 7105181487.897996
> INFO:__main__:Objective constant: 142949224.50470072
> ```
> 
> Results from main default.
> 
> ```shell
> INFO:__main__:Objective function: 7993595361.0
> INFO:__main__:Objective constant: 142949224.50470072
> ```
> 
> The percentage difference is about 1%.

Hi @ekatef, I have updated the objective result from the main run.

ekatef commented:
> > Just adding here the results from the linopy run
> > ```shell
> > INFO:__main__:Objective function: 7105181487.897996
> > INFO:__main__:Objective constant: 142949224.50470072
> > ```
>>
> > Results from main default.
> > ```shell
> > INFO:__main__:Objective function: 7993595361.0
> > INFO:__main__:Objective constant: 142949224.50470072
> > ```  
> >   
> > The percentage difference is about 1%.
> 
> Hi @ekatef, I have updated the objective result from the main run.

Thanks a lot @GbotemiB!! Agree that the difference look good for the objective constant. However, it looks like the difference for the objective function is >10%, which is quite a lot. It would be great to investigate the reason of this difference and hopefully find a way to fix it.

GbotemiB committed:
merge main branch

GbotemiB committed:
update pypsa-earth

GbotemiB commented:
@ekatef 
Linopy Run with latest update from [Linopy PR in pypsa-earth](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796).

```bash
INFO:__main__:Objective function: 7982030165.971943
INFO:__main__:Objective constant: 142949224.50470072
```
Results from Main Run
```bash
INFO:__main__:Objective function: 7993595361.0
INFO:__main__:Objective constant: 142949224.50470072
```

The percentage difference is now around 0.1% which was far from the initial difference which was around 10%.

FabianHofmann commented:
This is awesome @GbotemiB, well done! what is the status then with the linopy integration overall? is it stable in both pypsa-earth and pypsa-earth-sec? If yes, we could potentially merge them before the merge.

GbotemiB commented:
@FabianHofmann, linopy in pypsa-earth is ready, @ekatef can confirm this. I can confirm that linopy integration in pypsa-earth-sec is ready also🤞🏼

FabianHofmann commented:
nice! @ekatef @davide-f, I am already running a dry run locally on the merge, it does not seem to be too much work. so, we can now either merge the linopy integration first, and I'll update everything on my end, or roll out the merge first and then port the linopy changes.  I am fine with both :)

ekatef commented:
> @ekatef Linopy Run with latest update from [Linopy PR in pypsa-earth](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796).
> 
> ```shell
> INFO:__main__:Objective function: 7982030165.971943
> INFO:__main__:Objective constant: 142949224.50470072
> ```
> 
> Results from Main Run
> 
> ```shell
> INFO:__main__:Objective function: 7993595361.0
> INFO:__main__:Objective constant: 142949224.50470072
> ```
> 
> The percentage difference is now around 0.1% which was far from the initial difference which was around 10%.

Hello @GbotemiB, just to be sure: that's the update for the run of linopy-integration PR in PyPSA-Earth-Sec? If that's the case, the result is really amazing 🎉

ekatef commented:
> nice! @ekatef @davide-f, I am already running a dry run locally on the merge, it does not seem to be too much work. so, we can now either merge the linopy integration first, and I'll update everything on my end, or roll out the merge first and then port the linopy changes. I am fine with both :)

Hey @FabianHofmann, thank a lot for taking care about testing the merge strategy! 

I confirm that linopy integration in PyPSA-Earth is basically ready. There are still [some issues to be tracked](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796#issuecomment-2286545655). They don't look like blocking ones for me (and even do not relate to linopy integration mostly), but it would be perfect to have a view on that from @davide-f, once he'll be back to work in the beginning of the next week. That can be also helpful to align on the long-term development goals.

I'd also keep in mind the need to update the documentation to add some instructions on how to keep the model version stable. That's something that should be done in any case, and thanks to the recent work of @yerbol-akhmetov we still have some testing on this and the example of good practices. Just some announcements are needed, once we'll decide to launch the merge live 😄 

So, my feeling is that technically we are ready to any merge path, and happy to align on the details tomorrow

GbotemiB commented:
> > @ekatef Linopy Run with latest update from [Linopy PR in pypsa-earth](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796).
> > ```shell
> > INFO:__main__:Objective function: 7982030165.971943
> > INFO:__main__:Objective constant: 142949224.50470072
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Results from Main Run
> > ```shell
> > INFO:__main__:Objective function: 7993595361.0
> > INFO:__main__:Objective constant: 142949224.50470072
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The percentage difference is now around 0.1% which was far from the initial difference which was around 10%.
> 
> Hello @GbotemiB, just to be sure: that's the update for the run of linopy-integration PR in PyPSA-Earth-Sec? If that's the case, the result is really amazing 🎉

Yeah. The update is for the run of linopy-integration PR in PySA-Earth-Sec.

ekatef commented:
> > > @ekatef Linopy Run with latest update from [Linopy PR in pypsa-earth](https://github.com/pypsa-meets-earth/pypsa-earth/pull/796).
> > > ```shell
> > > INFO:__main__:Objective function: 7982030165.971943
> > > INFO:__main__:Objective constant: 142949224.50470072
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Results from Main Run
> > > ```shell
> > > INFO:__main__:Objective function: 7993595361.0
> > > INFO:__main__:Objective constant: 142949224.50470072
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The percentage difference is now around 0.1% which was far from the initial difference which was around 10%.
> > 
> > 
> > Hello @GbotemiB, just to be sure: that's the update for the run of linopy-integration PR in PyPSA-Earth-Sec? If that's the case, the result is really amazing 🎉
> 
> Yeah. The update is for the run of linopy-integration PR in PySA-Earth-Sec.

Brilliant :D Great work!

I think it could make sense to cross-check that we have answered to all the review comments from @davide-f  and @FabianHofmann (that is quite a big work, and that's quite easy to loose the overview...), and we may be close to complete. Awesome 😄

GbotemiB commented:
Closing this in favour of https://github.com/pypsa-meets-earth/pypsa-earth/pull/1182
